TAO/TAO_IDL/be_include/be_helper.h

```diff
@@ -167,6 +172,22 @@ class TAO_OutStream
   int gen_endif (void);
 
   // =overloaded operators
+#if defined (ACE_HAS_CPP11)
+  // Avoid duplication of overloaded operator for unsigned long
+  template <typename Dummy = TAO_OutStream &>
+  typename std::enable_if<std::is_same<Dummy, TAO_OutStream &>::value &&
+                          !std::is_same<ACE_CDR::ULongLong, unsigned long>::value &&
+                          !std::is_same<ACE_CDR::ULong, unsigned long>::value,
+                          TAO_OutStream &>::type
+  operator << (const unsigned long num)
+  {
+    ACE_OS::fprintf (this->fp_,
+                     "%lu",
+                     num);
```

Comment:
Will %lu always work? We have format specifiers in ace/Basic_Types.h which are dependent on the size of the type, for example ACE_UINT32_FORMAT_SPECIFIER_ASCII and ACE_UINT64_FORMAT_SPECIFIER_ASCII

Reply:
Are these changes in tao_idl still required now that there are no changes anymore in ace/Basic_Types.h?

Reply:
Yes, `%lu` should work since the input argument is `unsigned long` exactly. 

These changes are needed because there are functions that write variables of type `ACE_CString::size_type`, which maps to `unsigned long` on some platforms, to `TAO_OutStream`, and there is no `operator<<` for `unsigned long` from the existing code on such platforms.

Reply:
When `ACE_CString::size_type` is the problem, shouldn't we have a operator<< for `size_t` and use `ACE_SIZE_T_FORMAT_SPECIFIER_ASCII` as format specifier?

Reply:
That seems a better way to do it. I'll make those changes.

Reply:
So size_t is a distinct type (see my stackoverflow link below), thinking about it, shouldn't we have a `ACE_HAS_DISTINCT_SIZE_T` or something like that, so that in that case we enable the size_t overload, has it really something todo with the size of long/long long?

Reply:
Or the calling code has to convert to a known fixed size

Reply:
The underlying type of `size_t` is still some primitive type. For example, it's implemented as `unsigned long` in my macOS and also in the approved answer from the stackoverflow link. The problem is that the existing set of overloaded operators may not have an operator for the underlying type of `size_t` after `int64_t` and `uint64_t` are enabled. And that could possibly happen to other platforms as well, not just macOS. 

If the calling code can convert to a fixed-size type whose overloaded operator exists, then surely we won't need those changes. I think you, Adam, and Justin have a better idea of which design we should go for.

Reply:
At the idl level we normally work with types of a known size, with size_t this looks risky, a length of a string is 32bit in IDL, with a 64bit size_t it could cause issues, shouldnt the caller use CDR::ULong?

Reply:
There is some code in TAO_IDL (in be/be_visitor_operation/ami_cs.cpp if I remember correctly) that calls output operator with a variable of type `ACE_CString::size_type`, which is `size_t`.

Reply:
Shouldn't the caller there not cast first to a CORBA::ULong, the type of the length of a CORBA string, that should always be 32bit, never 64bit, which size_t could be

Reply:
Maybe the caller should do that. But I'm not experienced enough with ACE and TAO to tell which way the caller should do. That's the code I got to start with.

Reply:
What is the exact code where this is problematic? Can you add a link to it?

Reply:
Line 198 in ami_cs.cpp caused the error: [https://github.com/DOCGroup/ACE_TAO/blob/master/TAO/TAO_IDL/be/be_visitor_operation/ami_cs.cpp#L198](https://github.com/DOCGroup/ACE_TAO/blob/master/TAO/TAO_IDL/be/be_visitor_operation/ami_cs.cpp#L198).

Reply:
Since it's in TAO_IDL_BE, the ORB isn't directly seeing a size_t.  Instead, the code generator gets a size_t and writes it to the file as a C++ integer literal.  The type of this literal is `int` (for all reasonable string sizes), which is what will be seen by the ORB.

Reply:
Ok, not a problem with this addition